imodels/rule_set/slipper.py

```diff
@@ -0,0 +1,282 @@
+
+            if self._prune_rule_obj(X, y, rule) > self._prune_rule_obj(X, y, rule):
+                curr_rule = best_prune.copy()
+            else:
+                stop_condition = True
+
+        return curr_rule
+
+    def _pop_condition(self, rule, condition):
+        """
+        Remove a condition from an existing Rule object
+        """
+        temp = rule.copy()
+        temp.remove(condition)
+        return temp
+
+    def _make_default_rule(self, X, y):
+        """
+        Make the default rule that is true for every observation 
+        of data set. Without default rule a SlipperRule would never 
+        predict negative
+        """
+        default_rule = []
+        features = random.choices(
+            range(X.shape[1]),
+            k=random.randint(2, 8)
+        )
+
+        default_rule.append({
+            'feature': str(features[0]),
+            'operator': '>',
+            'pivot': str(min(X[:, features[0]]))
+        })
+
+        for i, x in enumerate(features):
+            if i % 2:
+                default_rule.append({
+                    'feature': x,
+                    'operator': '<',
+                    'pivot': str(max(X[:, x]))
+                })
+            else:
+                default_rule.append({
+                    'feature': x,
+                    'operator': '>',
+                    'pivot': str(min(X[:, x]))
+                })
+
+        return default_rule
+
+    def _prune_rule_obj(self, X, y, rule):
+        """
+        objective function for prune rule routine
+        eq 7 from Cohen & Singer (1999)
+        """
+
+        V_plus, V_minus = self._get_design_matrices(X, y, rule)
+        C_R = self.sample_weight(V_plus, V_minus)
+        return (1 - V_plus - V_minus) + V_plus * np.exp(-C_R) \
+            + V_minus * np.exp(C_R)
+
+    def _eq_5(self, X, y, rule):
+        """
+        equation 5 from Cohen & Singer (1999)
+        used to compare the learned rule with a default rule
+        """
+        W_plus, W_minus = self._get_design_matrices(X, y, rule)
+        return 1 - np.square(np.sqrt(W_plus) - np.sqrt(W_minus))
+
+    def _set_rule_or_default(self, X, y, learned_rule):
+        """
+        Compare output of eq 5 between learned rule and default rule
+        return rule that minmizes eq 5
+        """
+
+        rules = [self._make_default_rule(X, y), learned_rule]
+        scores = [self._eq_5(X, y, rule) for rule in rules]
+        self.rule = rules[scores.index(min(scores))]
+
+    def make_feature_dict(self, num_features, features):
+        """
+        Map features to place holder names
+        """
+        if features is None:
+            new_feats = ['feature_' + str(i) for i in range(num_features)]
+        else:
+            new_feats = features
+
+        self.feature_dict = {
+            old_feat:new_feat for old_feat, new_feat in enumerate(new_feats)
+        }
+
+    def predict_proba(self, X):
+
+        proba = self.predict(X)
```

Comment:
Not entirely sure what predict proba should return in the case of slipper. This just returns array of shape (n,2) where column 1 is raw predictions.

Reply:
I think this is good!